feat: Create URIs for Files ADR (#949)
edgexfoundry/edgex-docs: docs_src/design/adr/0027-URIs for Files.md

### Authentication 

#### basic-auth in URI

While not recommended, users will be able to specify **basic-auth** (`<username>:<password>@`) in the URI in plain text. While this is ok network wise when using HTTPS, it isn't good practice to have these credentials specified in configuration or other service files where the URI is specified.

!!! example - "Example plain text `basic-auth` in URI located in configuration"
    ```toml
    [UoM]
    UoMFile = "https://myuser:mypassword@example.com/uom.yaml"
    ```

#### Secure Credentials

In order to provide a secure way for users to specify credentials, the `edgexSecretName` query parameter can be specified on the URI. This parameter specifies a Secret Name from the service's Secret Store where the credentials reside and will be processed by the new helper function.

!!! example - "Example URI with `edgexSecretName` query parameter"
    ```toml
    [UoM]
    UoMFile = "https://example.com/uom.yaml?edgexSecretName=mySecretName"
    ```    

> The type of authentication as well as the credentials will be contained in the secret data specified by the Secret Name. Three types of authentication will be supported, which are `usernamepassword`, `apikey` and `bearertoken`. 

Reviewer: ```suggestion
The type of authentication as well as the credentials will be contained in the secret data specified by the Secret Name. One types of authentication will be supported: `customheader`, which will accommodate various forms of an `Authorization` header as well as typical API key usages. 

```

Reviewer 2: reworded

- When `usernamepassword` is specified as the type in the secret data, the **basic-auth** `username:password@` will be inserted into the URI using the **username** and **password** found in the secret data.

>     !!! example - "Example secret data - `usernamepassword`"
>         ```
>         type=usernamepassword
>         username=myuser
>         password=mypassword
>         ```
>     !!! example  - "Example resulting URI with `basic-auth`"
>         ```
>         https://myuser:mypassword@example.com/uom.yaml
>         ```

Reviewer: How about

```suggestion
        ```
        type=customheader
        headername=Authorization
        headercontents=Basic bXl1c2VyOm15cGFzc3dvcmQ=
        ```
        For a request header set as:
        ```
        GET https://example.com/uom.yaml HTTP/1.1
        Authorization: Basic bXl1c2VyOm15cGFzc3dvcmQ=
        ```
```

Reviewer 2: OK, @bnevis-i , what you are proposing is a more general form allowing for more then `Bearer`?

Reviewer: @lenny-intel Right.  All of the examples have a static http header and a static value -- just implement that and save some code.

Reviewer 2: replace `token` with you `customheader` and remove `apikey`

- When `apikey` is specified as the type in the secret data,  the **API Key** will be placed in the HTTP header using `apikeyname` and `apikeyvalue`  from the secret data

>     !!! example - "Example secret data - `apikey`"
>         ```
>         type=apikey
>         apikeyvalue=mykeyvalue
>         apikeyname=myname
>         ```

Reviewer: Let me try to do even better than that:

I find the example very confusing. Took me some time to realize what goes where.

```suggestion
    !!! example - "Example secret data - `customheader`"
        ```
        type=customheader
        headername=X-API-Key
        headercontents=secret-value-nobody-should-know
        ```
        For a request header set as:
        ```
        GET https://example.com/uom.yaml HTTP/1.1
        X-API-Key: secret-value-nobody-should-know
        ```
```

>         ```

Reviewer 3: I like your suggestion and prefer `headername` and `headervalue` to be consistent with HTTP spec terminology.

I was trying to keep the [`in=header|query` idea](https://github.com/edgexfoundry/edgex-docs/pull/949/files#r1097478177) open, which was rejected.

>         ```

Reviewer: @farshidtz It might be worthwhile to figure out how you could do it, technically.  I just suggest not implementing it unless it is absolutely needed.

>         ```

Reviewer 2: How is API Key non-standard?
Nope, we didn't have it specifically specify it in the UCR, but that doesn't mean a design can't include it.

>         ```

Reviewer 2: Yea, the in query option should be included.

>         ```

Reviewer 3: > How is API Key non-standard? 

It is some custom format whereas Authorization: Basic and Bearer are standardized i.e. specified by IETF which is a standardisation body.

> Nope, we didn't have it specifically specify it in the UCR, but that doesn't mean a design can't include it.

That's actually one of the reasons we do have UCRs, to decide on what exactly is required for what use case before diving into design. Otherwise, we can add more and more features without anyone actually needing or using them.

>         ```

Reviewer 2: @farshidtz , FYI I went with @bnevis-i approach and eliminated the specific API-key option

    
- When `bearertoken` is specified as the type in the secret data, the **Bearer Token** will be placed in the HTTP header as `Authorization` with value of `Bearer <token>`  where `token` is from the secret data. 

    !!! example - "Example secret data - `token`"
        ```
        type=bearertoken
        token=mytoken

>         ```

Reviewer: IMO this doesn't offer anything additional over "customheader"

- **Core Metadata's** loading of the **UOM file** will be adjusted to use the new file load function.

- **Device Service's** loading of **device profiles**, **device definitions** and **provision watchers** files will be adjusted to load an index file specified by a URI in place of the configured folder name. The contents of the index file will be used to load the individual files by URI  by appending the filenames to the original URI. Any authentication specified in the original URI will be used in the subsequent URIs. 

    !!! example - "Example DevicesDir configuration in service configuration" 
        ```toml
        [Device]
          ...
          ProfilesDir = "./res/profiles"
          DevicesDir = "http://example.com/devices/index.json"
          ProvisionWatchersDir = "./res/provisionwatchers"
          ...
        ```
    !!! example - "Example Device Index file `http://example.com/devices/index.json`"
        ```json
        [

>             "device1.yaml", "device2.yaml"

Reviewer 4: they should be device1.json and devic2.json for this example

        ]
        ```

    !!! example - "Example resulting device file URIs from above example"
        ```
        http://example.com/devices/device1.yaml

>         http://example.com/devices/device2.yaml

Reviewer 4: they should be device1.json and devic2.json for this example

Reviewer 2: Same response as above

    ```

#### Secure Credentials

In order to provide a secure way for users to specify credentials, the `edgexSecretName` query parameter can be specified on the URI. This parameter specifies a Secret Name from the service's Secret Store where the credentials reside and will be processed by the new helper function.

!!! example - "Example URI with `edgexSecretName` query parameter"
    ```toml
    [UoM]
    UoMFile = "https://example.com/uom.yaml?edgexSecretName=mySecretName"
    ```    

The type of authentication as well as the credentials will be contained in the secret data specified by the Secret Name. Two types of authentication will be supported, which are `usernamepassword`and `httpheader`. The `httpheader`type  will accommodate various forms of an `Authorization` header as well as typical API key usages

- When `usernamepassword` is specified as the type in the secret data, the `username:password@` will be inserted into the URI using the **username** and **password** found in the secret data.

    !!! example - "Example secret data - `usernamepassword`"
        ```
        type=usernamepassword
        username=myuser
        password=mypassword
        ```

>         Resulting URI with `basic-auth`"
>         ```
>         https://myuser:mypassword@example.com/uom.yaml

Reviewer 3: ```suggestion

Depending on the `WWW-Authenticate` header in the server's initial response, this results in Basic or Digest access authentication flows.
```

        ```
        
- When `httpheader` is specified as the type in the secret data, the header name and contents from the secret data  will be placed in the HTTP header. 

>     !!! example - "Example secret data - `Authorization` using  `httpheader`"

Reviewer 3: ```suggestion
    !!! example - "Example secret data - Basic access authentication using  `httpheader`"

```

        ```
        type=httpheader
        headername=Authorization
        headercontents=Basic bXl1c2VyOm15cGFzc3dvcmQ=
        ```
        For a request header set as:
        ```
        GET https://example.com/uom.yaml HTTP/1.1
        Authorization: Basic bXl1c2VyOm15cGFzc3dvcmQ=

>         ```

Reviewer 3: Given the changes to allow setting custom headers, these two mechanisms have become redundant. The credentials passed in the URI are just a common mechanism for when setting the header isn't possible for basic auth. The user can base64-encode the `user:pass` and set it as the `Authorization: Basic <base64-encoded>`.

As @bnevis-i suggested, one type is sufficient to cover all header-based authentication.

See what curl does with credentials passed in URL:
```
$ curl -v http://user:pass@example.com
*   Trying 93.184.216.34:80...
*   Trying 2606:2800:220:1:248:1893:25c8:1946:80...
* Connected to example.com (93.184.216.34) port 80 (#0)
* Server auth using Basic with user 'user'
> GET / HTTP/1.1
> Host: example.com
> Authorization: Basic dXNlcjpwYXNz  <------- The creds are sent to the server in header
> User-Agent: curl/7.81.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
...
```

Reviewer 3: With that said, I am fine with keeping raw username and password for basic auth for the sake of usability. But the type should be called specifically to indicate what it is: `basic-auth`. Moreover, we shouldn't put a basic auth example with custom header because we already have basic auth supported by accepting raw user/pass.

Reviewer: @farshidtz   I wanted to point out that a potential benefit of username password vs a raw header is that if the remote server requires digest auth, knowing the username and password would allow use of digest auth, whereas a raw header wouldn't.

Reviewer 3: Makes sense. But the example should NOT be:
```
        Resulting URI with `basic-auth`"
        ```
        https://myuser:mypassword@example.com/uom.yaml
        ```
```
Because in case of Basic Auth, the we should pass the given credentials in header as it gives lower chance of ending up in logs.

Also, the example in the case of `httpheader` should not show the basic auth, but rather a Bearer token example.

Reviewer 3: Based in this discussion, I've added suggestion below.

I kept the basic auth example for custom header, in case direct basic auth is needed without asking if server needs it.

Reviewer: > Makes sense. But the example should NOT be:

@farshidtz I agree.  Don't change the URI; just do the auth.  

> Also, the example in the case of httpheader should not show the basic auth, but rather a Bearer token example.

It should be an API key example.  Bearer tokens are supposed to have a short lifetime.  It's unlikely anyone interacting with a real system would have long-term success with a static bearer token.  Though technically it is possible.

Reviewer 3: Many servers use OAuth2 and in turn, need to pass a Bearer token. It is true that Bearer tokens should be short lived, but that doesn't prevent an application from refreshing that short-lived token in the EdgeX Secret Store to allow continuous access to the file server. Anyway, this is just an example and it would be technically possible regardless of the example.

Reviewer 2: @farshidtz , @bnevis-i and I discussed this and agreed that `httpheader` covers all needs including basic-auth. We will drop `usernamepassword` and only have `httpheader`. Examples will be updated to show `api-key`, `basic-auth`, and `Bearer`. At this time digest will not be supported. I can be added at a later time base on feedback of its need.

Reviewer 2: Updates have been pushed

>         

Reviewer 3: ```suggestion

    !!! example - "Example secret data - Bearer access authentication using `httpheader`"
        ```
        type=httpheader
        headername=Authorization
        headercontents=Bearer bXl0b2tlbg==
        ```
        For a request header set as:
        ```
        GET https://example.com/uom.yaml HTTP/1.1
        Authorization: Bearer bXl0b2tlbg==

```

!!! example - "Example plain text `username-password` in URI located in configuration"
    ```toml
    [UoM]
    UoMFile = "https://myuser:mypassword@example.com/uom.yaml"
    ```

#### Secure Credentials

In order to provide a secure way for users to specify credentials, the `edgexSecretName` query parameter can be specified on the URI. This parameter specifies a Secret Name from the service's Secret Store where the credentials reside and will be processed by the new helper function.

!!! example - "Example URI with `edgexSecretName` query parameter"
    ```toml
    [UoM]
    UoMFile = "https://example.com/uom.yaml?edgexSecretName=mySecretName"
    ```    

The type of authentication as well as the credentials will be contained in the secret data specified by the Secret Name. Only one type of authentication will be supported initially, which is `httpheader`. The `httpheader`type  will accommodate various forms of authorization placed in the header. Others types can be added in the future when need is determined.

!!! note

>     Digist Auth will not be supported at this time. It can be added in the future based on feedback indicating its need.

Reviewer 5: ```suggestion
    Digest Auth will not be supported at this time. It can be added in the future based on feedback indicating its need.

```

        Authorization: Bearer eyJhbGciO...
        ```

### Services/Files Impacted

- **All Services** will be impacted for enabling the loading the **common configuration** and **private configuration** files using URIs. This will be handled in `go-mod-bootstrap's` processing of the `-cc/--commonConfig` and `-cf/--configFile` command line flags.

- **Core Metadata's** loading of the **UOM file** will be adjusted to use the new file load function.

- **Device Service's** loading of **device profiles**, **device definitions** and **provision watchers** files will be adjusted to load an index file specified by a URI in place of the configured folder name. The contents of the index file will be used to load the individual files by URI  by appending the filenames to the original URI. Any authentication specified in the original URI will be used in the subsequent URIs. 

    !!! example - "Example DevicesDir configuration in service configuration" 
        ```toml
        [Device]
          ...
          ProfilesDir = "./res/profiles"
          DevicesDir = "http://example.com/devices/index.json"
          ProvisionWatchersDir = "./res/provisionwatchers"
          ...

>         ```

Reviewer 5: ```suggestion
        ```

```

Reviewer 5: Otherwise, second example doesn't start on its own line.

Reviewer 2: Added white space.

# URIs for Files ADR
### Submitters
- Lenny Goodell (Intel)

## Change Log
- [pending ](https://github.com/edgexfoundry/edgex-docs/pull/949) (2023-02-01)

## Referenced Use Case(s)
- [URIs for Files UCR](https://docs.edgexfoundry.org/3.0/design/ucr/URIs-for-Files/)

## Context

> Currently, in the Levski and earlier releases services can only load configuration, units of measurements, device profiles, device definitions, provision watches, etc. from the local file system. As outline in the reference UCR, there is a need to be able to load files from a remote locations using a URIs to specify the locations. 

Reviewer 5: ```suggestion
Currently, in the Levski and earlier releases, services can only load configuration, units of measurements, device profiles, device definitions, provision watches, etc. from the local file system. As outlined in the reference UCR, there is a need to be able to load files from remote locations using URIs to specify the locations. 

```

Reviewer 2: fixed

## Proposed Design
This ADR proposes a new helper function for loading files be added to `go-mod-bootstrap`. This function will provide the logic for loading a file either from local file system (as is today) or from a remote location. As stated in the UCR, only **HTTP** and **HTTPS** URIs will be supported. For **HTTPS**, certificate validation will be performed using the system's built-in trust anchors. The docker images for all services will have the CA certs installed as is done here in App Service Configurable's [Dockerfile](https://github.com/edgexfoundry/app-service-configurable/blob/v2.3.0/Dockerfile#L46).

### Authentication 

> While not recommended, users will be able to specify **basic-auth** (`<username>:<password>@`) in the URI in plain text. In order to provide a secure way for users to specify credentials, the URI scheme will be expanded to allow specifying a Secret Name from the service's Secret Store. This will take the form  `[secretName: <secret-name>]@`. The type of authentication as well as the credentials will be contained in the secret data specified by the Secret Name. Three types of authentication will be supported, which are `usernamepassword`, `apikey` and `bearertoken`. 

Reviewer 5: There's a space between secretName: and <secret-name> - is that valid in a URL, even if inside the square brackets?
The square brackets here are literal and don't have the usual meaning of "optional parameter", right?

Reviewer 2: Yes, the space probably should be removed.
`[secretName: <secret-name>]@` is our own addition and will always be replaced or removed so will never be in the actual URL that is used.

Reviewer 2: Removed space

> - When `usernamepassword` is specified the `[secretName: <secret-name>]@` text in the URI will be replaced with `username:password@` using the **username** and **password** found in the secret data.

Reviewer 5: ```suggestion
- When `usernamepassword` is specified at the type in the Secret Store, the `[secretName: <secret-name>]@` text in the URI will be replaced with `username:password@` using the **username** and **password** found in the secret data.

```

Reviewer 2: Added

        password=mypassword
        ```

- When `apikey` is specified, the `[secretName: <secret-name>]@` text will be removed from the URI and the **API Key** will be placed in the HTTP header using `apikeyname` and `apikeyvalue`  from the secret data

    !!! example - "Example secret data - `apikey`"
        ```
        type=apikey
        apikeyvalue=mykeyvalue
        apikeyname=myname
        ```
    
- When `bearertoken` is specified, the `[secretName: <secret-name>]@` text will be removed from the URI and the **Bearer Token** will be placed in the HTTP header as  `Authorization` with value of `Bearer <token>`  where `token` is from the secret data. 

    !!! example - "Example secret data - `token`"
        ```
        type=bearertoken
        token=mytoken
        ```

### Services/Files Impacted

- **All Services** will be impacted for enabling the loading the **common configuration** and **private configuration** files using URIs. This will be handled in `go-mod-bootstrap's` processing of the `-cc/--commonConfig` and `-cf/--configFile` command line flags.

- **Core Metadata's** loading of the **UOM file** will be adjusted to use the new file load function.

> - **Device Service's** loading of **device profiles**, **device definitions** and **provision watches** files will be adjusted to load an index file specified by a URI in place of the configured folder name. The contents of the index file will be used to load the individual files by URI  by appending the filenames to the original URI. Any authentication specified in the original URI will be used in the subsequent URIs. 

Reviewer 6: **provision watchers**

Reviewer 2: fixed

    !!! example - "Example Device Index file http://example.com/devices/index.json"
        ```json
        [
            "device1.yaml", "device2.yaml"
        ]

>         ```

Reviewer 5: This example needs to be completed by showing the resulting URI. I'm not getting the picture.

Reviewer 2: Ok

Reviewer 2: added

    !!! example - "Example secret data - `token`"
        ```
        type=bearertoken
        token=mytoken
        ```

### Services/Files Impacted

- **All Services** will be impacted for enabling the loading the **common configuration** and **private configuration** files using URIs. This will be handled in `go-mod-bootstrap's` processing of the `-cc/--commonConfig` and `-cf/--configFile` command line flags.

- **Core Metadata's** loading of the **UOM file** will be adjusted to use the new file load function.

- **Device Service's** loading of **device profiles**, **device definitions** and **provision watches** files will be adjusted to load an index file specified by a URI in place of the configured folder name. The contents of the index file will be used to load the individual files by URI  by appending the filenames to the original URI. Any authentication specified in the original URI will be used in the subsequent URIs. 

    !!! example - "Example Device Index file `http://example.com/devices/index.json`"
        ```json
        [
            "device1.yaml", "device2.yaml"
        ]
        ```

>     !!! example - "Example resulting device file URIs from above example"

Reviewer 5: This is better, but I still had to read it 5 times to see what you meant. Maybe to complete the example, show a device definition that is set to 'http://example.com/devices/index.json' to result in these URIs.
I wasn't getting where 'http://example.com/devices/' was introduced into the example pictured.

Reviewer 2: Added example of the DeviceDir configuration
